### PR TITLE
fix: write object to the correct urn when moving from another storage to object store

### PR DESCRIPTION
**Pull request by icewind1991**
The default `copyFromStorage` + `delete` approach loads to the object being written to the wrong urn.

See also https://github.com/nextcloud/server/pull/43660 for a description of the problem and alternate approach.

### CODE REVIEW

File: lib/private/Files/ObjectStore/ObjectStoreStorage.php [hunk 1]
```diff
@@ -594,6 +594,30 @@ public function copyFromStorage(
...
+				$this->moveFromStorage($sourceStorage, $child->getPath(), $targetInternalPath . '/' . $child->getName());
+			}
+			$sourceStorage->rmdir($sourceInternalPath);
+		} else {
+			// move the cache entry before the contents so that we have the correct fileid/urn for the target
+			$this->getCache()->moveFromCache($sourceCache, $sourceInternalPath, $targetInternalPath);
+			try {
+				$this->writeStream($targetInternalPath, $sourceStorage->fopen($sourceInternalPath, 'r'), $sourceCacheEntry->getSize());
+			} catch (\Exception $e) {
+				$this->getCache()->remove($targetInternalPath);
```
Comment (come-nc) [inline lib/private/Files/ObjectStore/ObjectStoreStorage.php:613]: Should we also put back the file in the source cache on exceptions?
Here it’s lost.

Comment (icewind1991) [reply lib/private/Files/ObjectStore/ObjectStoreStorage.php]: done

**Review by artonge — APPROVED** [review]

**Review by kesselb — APPROVED** [review]
I've run into the problem, that moving a file from S3 external to S3 primary uses the wrong urn, on my development setup. The patch resolves the problem.

**Comment by joshtrichards** [comment]
On the surface, this appears like it *could* (untested) also fix: #31870 / #25693 (which was attempted in #32781).

**Pull request merged**